gitlab-scm-filter-aged-refs/src/main/java/org/jenkinsci/plugins/scm_filter/GitLabAgedRefsTrait.java

```diff
@@ -0,0 +1,111 @@
+import jenkins.scm.api.trait.SCMSourceContext;
+import jenkins.scm.api.trait.SCMSourceRequest;
+import jenkins.scm.impl.trait.Selection;
+import org.gitlab4j.api.GitLabApi;
+import org.gitlab4j.api.GitLabApiException;
+import org.gitlab4j.api.models.Branch;
+import org.gitlab4j.api.models.Commit;
+import org.gitlab4j.api.models.MergeRequest;
+import org.jenkinsci.Symbol;
+import org.kohsuke.stapler.DataBoundConstructor;
+
+public class GitLabAgedRefsTrait extends AgedRefsTrait {
+    public static final Logger LOGGER = Logger.getLogger(GitLabAgedRefsTrait.class.getName());
+
+    /**
+     * Constructor for stapler.
+     *
+     * @param retentionDays retention period in days
+     */
+    @DataBoundConstructor
+    public GitLabAgedRefsTrait(String retentionDays) {
+        super(retentionDays);
+    }
+
+    @Override
+    protected void decorateContext(SCMSourceContext<?, ?> context) {
+        if (retentionDays > 0) {
+            context.withFilter(new ExcludeOldBranchesSCMHeadFilter(retentionDays));
+        }
+    }
+
+    /**
+     * Our descriptor.
+     */
+    @Extension
+    @Selection
+    @Symbol("gitLabAgedRefsTrait")
+    @SuppressWarnings("unused") // instantiated by Jenkins
+    public static class DescriptorImpl extends AgedRefsDescriptorImpl {
+
+        @Override
+        public Class<? extends SCMSourceContext> getContextClass() {
+            return GitLabSCMSourceContext.class;
+        }
+
+        @Override
+        public Class<? extends SCMSource> getSourceClass() {
+            return GitLabSCMSource.class;
+        }
+    }
+
+    /**
+     * Filter that excludes references (branches, pull requests, tags) according to their last commit modification date and the defined retentionDays.
+     */
+    private static class ExcludeOldBranchesSCMHeadFilter extends ExcludeBranchesSCMHeadFilter {
+
+        ExcludeOldBranchesSCMHeadFilter(int retentionDays) {
+            super(retentionDays);
+        }
+
+        @Override
+        public boolean isExcluded(@NonNull SCMSourceRequest scmSourceRequest, @NonNull SCMHead scmHead)
+                throws IOException {
+            if (scmHead instanceof BranchSCMHead) {
+                Iterable<Branch> branches = ((GitLabSCMSourceRequest) scmSourceRequest).getBranches();
+                for (Branch branch : branches) {
+                    if (branch.getName().equals(scmHead.getName())) {
+                        long branchTS = branch.getCommit().getCommittedDate().getTime();
+                        return branchTS < getAcceptableDateTimeThreshold();
+                    }
+                }
+            } else if (scmHead instanceof MergeRequestSCMHead) {
+                MergeRequestSCMHead mrHead = (MergeRequestSCMHead) scmHead;
+                GitLabSCMSourceRequest gitLabSCMSourceRequest = (GitLabSCMSourceRequest) scmSourceRequest;
+                Iterable<MergeRequest> mrs = gitLabSCMSourceRequest.getMergeRequests();
+                for (MergeRequest mr : mrs) {
```

Comment:
For MR you can just use the original branch name from the MergeRequestSCMHead and iterate of the branches like you did with BranchSCMHead.
Not needed to make an additional api call.
```java
 String branchName = scmHead.getName();
 if (head instanceof MergeRequestSCMHead mrHead) {
   branchName = mrHead.getOriginName();
 }
 for (Branch branch : branches) {
     if (branch.getName().equals(branchName )) {
       ....
     }
 }
```
I implemented the same concept in gitlab-branch-source plugin but I was directed to this implementation which does the same think I did in the gitlab-branch-source pluigin.